### PR TITLE
Move MPI behavior from HTEX to MPIExecutor

### PR DESCRIPTION
**Pull request by yadudoc**
# Description

This PR moves the following MPI related functionality and options from HTEX to MPIExecutor:

* Kwarg options `enable_mpi_mode` and `mpi_launcher` is now removed from HTEX
* Checks for launcher being set to SimpleLauncher
* Checks for a valid `mpi_launcher` in now in MPIExecutor
* A new `validate_resource_specification` method is added to HTEX that currently asserts that no resource_specification is passed to it, since HTEX does not support any such options
* MPIExecutor overrides `validate_resource_specification` to check for a valid MPI resource specification

These changes should make it easier to have executor specific resource validation.

# Changed Behaviour

* HTEX kwarg `enable_mpi_mode` and `mpi_launcher` are no longer supported.
* Expect to use MPI functionality *only* through the MPIExecutor

## Type of change

I'm marking this as maintenance since the docs have moved on to only advertize MPI functionality through MPIExecutor.

- Code maintenance/cleanup

### CODE REVIEW

**Review by benclifford — CHANGES_REQUESTED** [review]

File: parsl/executors/high_throughput/executor.py [hunk 1]
```diff
@@ -348,6 +324,8 @@ def __init__(self,
         self.interchange_launch_cmd = interchange_launch_cmd
 
     radio_mode = "htex"
+    enable_mpi_mode: bool = False
+    mpi_launcher: str = "mpiexec"
```
[the remark this answers is not in the document]
Comment (yadudoc) [reply parsl/executors/high_throughput/executor.py:328]: I agree that it is weird. The issue I see is that the manager, and worker also currently expect a string for mpi_launcher that they simply ignore when not in mpi_mode. I can fix this as a separate PR, what do you think?

[the remark this answers is not in the document]
Comment (benclifford) [reply parsl/executors/high_throughput/executor.py:328]: yeah i think there is a bunch of different structuring that could happen more fundamentally in the worker pool code - which could take away all MPI references from the base htex entirely.

File: parsl/executors/high_throughput/mpi_executor.py [hunk 2]
```diff
@@ -30,6 +35,11 @@ class MPIExecutor(HighThroughputExecutor):
     max_workers_per_block: int
         Maximum number of MPI applications to run at once per block
 
+    mpi_launcher: str
+        This field is only used if enable_mpi_mode is set. Select one from the
```
[the remark this answers is not in the document]
Comment (yadudoc) [reply parsl/executors/high_throughput/mpi_executor.py]: Fixed.

File: parsl/executors/high_throughput/executor.py [hunk 3]
```diff
@@ -377,6 +355,17 @@ def worker_logdir(self):
             return "{}/{}".format(self.worker_logdir_root, self.label)
         return self.logdir
 
+    def validate_resource_spec(self, resource_specification: dict):
+        """HTEX does not support *any* resource_specification options and
+        will raise InvalidResourceSpecification is any are passed to it"""
+        if resource_specification:
+            raise InvalidResourceSpecification(
+                set(resource_specification.keys()),
+                ("HTEX does not support MPI applications, please use the "
```
[the remark this answers is not in the document]
Comment (yadudoc) [reply parsl/executors/high_throughput/executor.py]: I've updated the wording here to point to MPIExecutor and WorkQueueExecutor.

File: parsl/executors/high_throughput/mpi_executor.py [hunk 4]
```diff
@@ -86,5 +96,16 @@ def __init__(self,
             block_error_handler=block_error_handler,
             encrypted=encrypted
         )
+        self.enable_mpi_mode = True
 
         self.max_workers_per_block = max_workers_per_block
+
+        assert isinstance(self.provider.launcher, parsl.launchers.SimpleLauncher), \
```
[the remark this answers is not in the document]
Comment (yadudoc) [reply parsl/executors/high_throughput/mpi_executor.py]: Fixed.

File: parsl/executors/high_throughput/executor.py [hunk 5]
```diff
@@ -25,8 +24,7 @@
     RandomManagerSelector,
 )
 from parsl.executors.high_throughput.mpi_prefix_composer import (
-    VALID_LAUNCHERS,
-    validate_resource_spec,
+    InvalidResourceSpecification,
```
Comment (benclifford) [inline parsl/executors/high_throughput/executor.py:27]: this is weird code placement. I made issue #3589

File: parsl/tests/test_mpi_apps/test_bad_mpi_config.py [hunk 6]
```diff
@@ -1,33 +1,49 @@
...
+                MPIExecutor(
                     provider=SlurmProvider(launcher=launcher),
                 )
             ])
 
 
 @pytest.mark.local
-def test_correct_launcher_with_mpi_mode():
+def test_bad_mpi_launcher():
+    """ValueError if an unsupported mpi_launcher is specified"""
```
Comment (benclifford) [inline parsl/tests/test_mpi_apps/test_bad_mpi_config.py:24]: this test looks for a type error not a value error?

File: test-requirements.txt [hunk 7]
```diff
@@ -7,11 +7,13 @@ pytest-random-order
 nbsphinx
 sphinx_rtd_theme
 mypy==1.5.1
+types-mock
 types-python-dateutil
 types-requests
 types-paramiko
 mpi4py
 
+
```
Comment (benclifford) [inline test-requirements.txt:17]: might as well tidy this up when addressing the value error comment

**Review by benclifford — CHANGES_REQUESTED** [review]

File: parsl/tests/test_mpi_apps/test_bad_mpi_config.py (hunk 6 above)
Comment (yadudoc) [reply parsl/tests/test_mpi_apps/test_bad_mpi_config.py]: Fixed.

File: test-requirements.txt (hunk 7 above)
Comment (yadudoc) [reply test-requirements.txt:17]: Fixed.

File: parsl/executors/high_throughput/executor.py (hunk 5 above)
Comment (yadudoc) [reply parsl/executors/high_throughput/executor.py:27]: I think this is better addressed in a separate PR. What do you think?

File: test-requirements.txt [hunk 8]
```diff
@@ -25,4 +25,3 @@ Sphinx==4.5.0
 twine
 wheel
 isort
-
```
Comment (benclifford) [inline test-requirements.txt:26]: are you trolling me

File: parsl/executors/high_throughput/executor.py (hunk 5 above)
Comment (benclifford) [reply parsl/executors/high_throughput/executor.py:27]: right, that's what I made the issue for

**Review by benclifford — APPROVED** [review]

**Pull request merged**